src/main/java/com/commercetools/project/sync/category/service/impl/CategoryReferenceTransformServiceImpl.java

```diff
@@ -0,0 +1,159 @@
+package com.commercetools.project.sync.category.service.impl;
+
+import static java.util.stream.Collectors.toList;
+import static org.apache.commons.lang3.StringUtils.isBlank;
+
+import com.commercetools.project.sync.category.service.CategoryReferenceTransformService;
+import com.commercetools.project.sync.model.ResourceIdsGraphQlRequest;
+import com.commercetools.project.sync.service.impl.BaseServiceImpl;
+import com.commercetools.project.sync.util.referenceresolution.CategoryReferenceResolutionUtils;
+import com.commercetools.sync.commons.models.GraphQlQueryResources;
+import com.commercetools.sync.commons.models.ResourceKeyId;
+import io.sphere.sdk.categories.Category;
+import io.sphere.sdk.categories.CategoryDraft;
+import io.sphere.sdk.client.SphereClient;
+import io.sphere.sdk.models.Asset;
+import io.sphere.sdk.models.Reference;
+import io.sphere.sdk.types.Type;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.stream.Collectors;
+import javax.annotation.Nonnull;
+
+public class CategoryReferenceTransformServiceImpl extends BaseServiceImpl
+    implements CategoryReferenceTransformService {
+
+  public CategoryReferenceTransformServiceImpl(@Nonnull final SphereClient ctpClient) {
+    super(ctpClient);
+  }
+
+  /**
+   * Given a {@link List} of categories, this method will transform Categories to CategoryDrafts by
+   * resolving the References parent Category, {@link Type}. If there exists a mapping for all the
+   * ids in {@code referenceIdToKeyCache}, the method replaces id field with key. If there is at
+   * least one missing mapping, it attempts to make a GraphQL request to CTP to fetch all ids and
+   * keys of every missing type. For each fetched key/id pair, the method will insert it into the
+   * {@code referenceIdToKeyCache} cache after the request is successful.
+   *
+   * @param categories the categories to find a key mapping for all the references and cache them.
+   * @return categoryDrafts with the references replaced Id with keys.
+   */
+  @Nonnull
+  @Override
+  public CompletionStage<List<CategoryDraft>> transformCategoryReferences(
```

Comment:
Would be better to add the javadocs for the interface contract instead of its implementation, wdyt?

Reply:
https://github.com/commercetools/commercetools-project-sync/pull/240/commits/add1685069142c9a57067e0dc11be9ba87589a60